# **Advanced DataBases Project**

**Συγγραφείς:** Γκαβογιάννης Κωνσταντίνος (el21820),
                Μεγαλιός Γεώργος (el21118)

**Ομάδα:** 10 \
**Ημερομηνία:** 11-2025 \
**Version:** 1.0  

---

## **Εισαγωγή**
Στο έργο αυτό μελετούμε δεδομένα εγκληματικότητας, δημογραφικά στοιχεία και οικονομικούς δείκτες για την περιοχή του Los Angeles, αξιοποιώντας τεχνικές επεξεργασίας μεγάλων δεδομένων.

## **Περιγραφή**
Στην παρούσα εξαμηνιαία εργασία ζητείται ανάλυση σε (μεγάλα) σύνολα δεδομένων, εφαρμόζοντας
επεξεργασία με τεχνικές που εφαρμόζονται σε data science projects. Τα εργαλεία που θα χρησιμοποιη-
θούν στα πλαίσια του project είναι τα Apache Hadoop (version>=3.0) και Apache Spark (version>=3.5).
Χρησιμοποιούμε τους πόρους στο ειδικά διαμορφωμένο περιβάλλον που σας έχει παρα-
χωρηθεί στο AWS cloud. Συνοπτικά, ο σκοπός της εργασίας είναι:
 - η εξοικείωση και ανάπτυξη των δεξιοτήτων των σπουδαστών στην εγκατάσταση και διαχείριση
των κατανεμημένων συστημάτων Apache Spark και Apache Hadoop.
 - Η χρήση σύγχρονων τεχνικών μέσω των API του Spark για την ανάλυση δεδομένων όγκου.
 - H κατανόηση των δυνατοτήτων και περιορισμών των εργαλείων αυτών σε σχέση με τους δια-
θέσιμους πόρους και τις ρυθμίσεις που έχουν επιλεγεί.

---
## **Δεδομένα**
Χρησιμοποιούμε σύνολα δεδομένων που είναι δημόσια διαθέσιμα, συγκεντρωμένα σε ενιαίο S3 bucket. Τα βασικά και δευτερεύοντα δεδομένα περιλαμβάνουν:

### **Τοποθεσία S3**
`s3://initial-notebook-data-bucket-dblab-905418150721/`

### **Σύνολα Δεδομένων και Πηγές**

- **Los Angeles Crime Data 2010–2019**  
  Πηγή: [Crime Data 2010–2019](https://data.lacity.org/Public-Safety/Crime-Data-from-2010-to-2019/63jg-8b9z)

- **Los Angeles Crime Data 2020–2025**  
  Πηγή: [Crime Data 2020–Σήμερα](https://data.lacity.org/Public-Safety/Crime-Data-from-2020-to-Present/2nrs-mtv8)

- **Census Blocks 2020**  
  Πηγή: [2020 Census Blocks](https://data.lacounty.gov/maps/lacounty::2020-census-blocks)

- **Census Blocks Fields**  
  Πηγή: [Data.gov Dataset Catalog](https://catalog.data.gov/dataset)

- **Median Household Income by ZIP Code (2021)**  
  Πηγή: [LA Almanac – Income 2021](http://www.laalmanac.com/employment/em12c_2021.php)

- **LA Police Stations**  
  Πηγή: [LAPD Police Stations](https://geohub.lacity.org/datasets/lahub::lapd-police-stations/explore)

- **Race & Ethnicity Codes**  

- **MO Codes**  
  Πηγή: [MO Codes](https://data.lacity.org/api/views/63jg-8b9z/files/e14442b9-a6b8-4531-83f3-f7ba980b1377)

  
Το βασικό Dataset αφορά καταγραφές εγκλημάτων από το 2010 μέχρι σήμερα. Τα υπόλοιπα σύνολα χρησιμοποιούνται συμπληρωματικά για χωρική, κοινωνική και οικονομική ανάλυση. Στις πήγες υπάρχουν όλα τα απαραίτητα μεταδεδομένα και η απαραίτητη τεκμιρίωση για την καλύτερη κατανόηση των Dataset.

---
## **Ερευνητικά Ερωτήματα**

1. **Ανάλυση Ηλικιακών Ομάδων Θυμάτων**  
   Ταξινομούμε σε φθίνουσα σειρά τις ηλικιακές ομάδες θυμάτων που σχετίζονται με περιστατικά “βαριάς σωματικής βλάβης”.  
   Ομάδες:  
   - Παιδιά: < 18  
   - Νεαροί ενήλικοι: 18–24  
   - Ενήλικοι: 25–64  
   - Ηλικιωμένοι: > 64  

2. **Φυλετικά Γκρουπ με τον Μεγαλύτερο Αριθμό Θυμάτων**  
   Ανά έτος βρίσκουμε τα τρία γκρουπ με τους περισσότερους καταγεγραμμένους παθόντες (Vict Descent) στο Los Angele και υπολογίζουμε το ποσοστό τους επί του συνόλου.

3. **Συχνότητες Μεθόδων Διάπραξης (MOcodes)**  
   Ταξινομούμε τις μεθόδους διάπραξης εγκλημάτων σε φθίνουσα σειρά συχνότητας, αντιστοιχίζοντας τους κωδικούς στις περιγραφές τους.

4. **Απόσταση από Αστυνομικά Τμήματα**  
   Υπολογίζουμε, για κάθε τμήμα, τον αριθμό περιστατικών που βρίσκονται πιο κοντά σε αυτό από κάθε άλλο, καθώς και τη μέση απόσταση.

5. **Συσχέτιση Εισοδήματος και Εγκληματικότητας**  
   Με βάση δεδομένα απογραφής (2020) και εισοδήματος (2021), υπολογίζουμε τη συσχέτιση μεταξύ κατακεφαλήν εισοδήματος και εγκληματικότητας ανά περιοχή.  
   Επαναλαμβάνουμε την ανάλυση για τις 10 περιοχές με το υψηλότερο και τις 10 με το χαμηλότερο εισόδημα.

Παρακάτω υπάρχουν αναλυτικές εκφωνήσεις των ερωτημάτων, κάθως και ο απαραίτητος κώδικας για την εκτέλεση τους.

---


## **Σταθερες και Βιβλιοθήκες για την χρήση στο Project.**
Σε αυτό το τμήμα δημιουργούνται βασικές σταθερές που θα χρησιμοποιηθούν σε όλο το Project. 


In [1]:
# Libraries and Utils
from pyspark.sql import SparkSession
from pyspark.sql.functions import row_number, substring, row_number, round, expr, regexp_replace, regexp_extract, split, col, when, udf, to_json, avg, radians, broadcast, count, sum as _sum, year, to_date
from pyspark.sql.types import StructField, StructType, IntegerType, FloatType, StringType, DoubleType, TimestampType
from pyspark.sql.window import Window
from sedona.register import SedonaRegistrator
from sedona.utils import SedonaKryoRegistrator, KryoSerializer
from sedona.spark import *
import time

# Constants
GROUP_NUMBER = 10
S3_OUR_GROUP_BUCKET_URL = "s3://groups-bucket-dblab-905418150721/group10/"

# Data Set S3 URLs for the Los Angeles Crime statistics
CRIME_DATA_2010_2019_CSV_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/LA_Crime_Data/LA_Crime_Data_2010_2019.csv"
CRIME_DATA_2020_2025_CSV_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/LA_Crime_Data/LA_Crime_Data_2020_2025.csv"
CENSUS_BLOCKS_2020_GEOJSON_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/LA_Census_Blocks_2020.geojson"
CENSUS_BLOCKS_FIELDS_CSV_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/LA_Census_Blocks_2020_fields.csv"
INCOME_BY_ZIP_2021_CSV_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/LA_income_2021.csv"
POLICE_STATIONS_CSV_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/LA_Police_Stations.csv"
RACE_ETHNICITY_CODES_CSV_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/RE_codes.csv"
MO_CODES_TXT_URL = "s3://initial-notebook-data-bucket-dblab-905418150721/project_data/MO_codes.txt"

# Schemas for the given Datasets in the Raw form found inside their respective CSVs
RAW_CRIME_SCHEMA = StructType([
    StructField("division_record_number", StringType(), False),         # DR_NO 
    StructField("report_date", StringType(), False),                    # DATE RPTD
    StructField("occurrence_date", StringType(), False),                # DATE OCC
    StructField("occurrence_time", IntegerType(), False),               # TIME OCC 
    StructField("geographic_area_id", IntegerType(), False),            # AREA 
    StructField("geographic_area_name", StringType(), False),           # AREA NAME
    StructField("reporting_district_number", IntegerType(), False),     # Rpt Dist No (4-digit code)
    StructField("part_1_or_2", IntegerType(), False),                   # Part 1-2
    StructField("crime_code", IntegerType(), True),                     # Crm Cd
    StructField("crime_code_description", StringType(), True),          # Crm Cd Desc
    StructField("mo_codes", IntegerType(), True),                       # Mocodes (stands for Modus Operandi)
    StructField("victim_age", IntegerType(), True),                     # Vict Age
    StructField("victim_sex", StringType(), True),                      # Vict Sex
    StructField("victim_descent_code", StringType(), True),             # Vict Descent
    StructField("premise_code", IntegerType(), True),                   # Premis Cd
    StructField("premise_description", StringType(), True),             # Premis Desc
    StructField("weapon_code", IntegerType(), True),                    # Weapon Used Cd
    StructField("weapon_description", StringType(), True),              # Weapon Desc
    StructField("case_status_code", StringType(), True),                # Status
    StructField("case_status_description", StringType(), True),         # Status Desc
    StructField("crime_code_1", IntegerType(), False),                  # Crm Cd 1
    StructField("crime_code_2", IntegerType(), True),                   # Crm Cd 2
    StructField("crime_code_3", IntegerType(), True),                   # Crm Cd 3
    StructField("crime_code_4", IntegerType(), True),                   # Crm Cd 4
    StructField("crime_location_address", StringType(), True),          # LOCATION
    StructField("crime_cross_street", StringType(), True),              # Cross Street
    StructField("latitude", DoubleType(), False),                       # LAT
    StructField("longitude", DoubleType(), False)                       # LON
])

CENSUS_BLOCKS_SCHEMA = StructType([                        # Field where found in the census_block_fields_csv
    StructField("object_id", StringType(), True),                   # OBJECTID – Object identifier (unique record ID)
    StructField("state_fips", StringType(), True),                  # State – State FIPS code (“06” = California)
    StructField("county_fips", StringType(), True),                 # COUNTY – County FIPS code (“037” = Los Angeles County)
    StructField("census_tract_2020", StringType(), True),           # CT20 – Census tract number (6 digits)
    StructField("block_group_2020", StringType(), True),            # BG20 – Block group number (7 digits)
    StructField("block_number_2020", StringType(), True),           # CB20 – Census block number (4 digits)
    StructField("tract_block_combined_2020", StringType(), True),   # CTCB20 – Combined tract + block identifier
    StructField("feature_type", StringType(), True),                # FEAT_TYPE – Feature type (land, water, ocean)
    StructField("county_fips_2020", StringType(), True),            # FIP20 – County FIPS (2020)
    StructField("blockgroup_fips_combined_2020", StringType(), True),# BGFIP20 – Block group + FIPS combined
    StructField("city_name", StringType(), True),                   # CITY – Incorporated city name
    StructField("community_name", StringType(), True),              # COMM – Unincorporated community or LA neighborhood
    StructField("city_community_label", StringType(), True),        # CITYCOMM – Combined city/community label
    StructField("zcta_2020", StringType(), True),                   # ZCTA20 – ZIP Code Tabulation Area
    StructField("health_district_2022", IntegerType(), True),       # HD22 – Health District number (2022)
    StructField("health_district_name", StringType(), True),        # HD_NAME – Health District name
    StructField("spa_2022", IntegerType(), True),                   # SPA22 – Service Planning Area number (2022)
    StructField("spa_name", StringType(), True),                    # SPA_NAME – Service Planning Area name
    StructField("supervisorial_district_2021", StringType(), True), # SUP21 – Supervisorial District number (2021)
    StructField("supervisorial_district_label", StringType(), True),# SUP_LABEL – Supervisorial District label
    StructField("population_2020", IntegerType(), True),            # POP20 – 2020 population count
    StructField("housing_units_2020", IntegerType(), True),         # HOUSING20 – Total housing units count
    StructField("fips_current", StringType(), True),                # FIP_CURRENT – Current FIPS code
    StructField("blockgroup_fips_current", StringType(), True),     # BG20FIP_CURRENT – Current block group + FIPS combination
    StructField("city_current", StringType(), True),                # CITY_CURRENT – Current city name
    StructField("community_current", StringType(), True),           # COMM_CURRENT – Current community name
    StructField("city_community_current", StringType(), True),      # CITYCOMM_CURRENT – Current combined city/community label
    StructField("geometry", StringType(), True),                    # Shape – Polygon geometry
    StructField("geometry_area", DoubleType(), True),               # Shape.STArea() – Area of census block polygon
    StructField("geometry_length", DoubleType(), True)              # Shape.STLength() – Perimeter length of census block polygon
])

RAW_INCOME_BY_ZIP_2021_SCHEMA = StructType([
    StructField("ZIP", IntegerType(), False),
    StructField("community", StringType(), False),
    StructField("median_income_str", StringType(), False)
])
INCOME_BY_ZIP_2021_SCHEMA = StructType([
    StructField("ZIP", IntegerType(), False),
    StructField("community", StringType(), False),
    StructField("median_income", IntegerType(), False)
])

POLICE_STATIONS_SCHEMA = StructType([
    StructField("longitude", DoubleType(), False),     # X
    StructField("latitude", DoubleType(), False),    # Y
    StructField("FID", IntegerType(), False),
    StructField("division", StringType(), False),
    StructField("location", StringType(), False),
    StructField("PDEC", IntegerType(), False)
])

RACE_ETHNICITY_CODES_SCHEMA = StructType([
    StructField("ID", StringType(), False),      # Vict Descent
    StructField("descent", StringType(), False)   # Vict Descent Full
])

MO_CODES_SCHEMA = StructType([
    StructField("mo_code", IntegerType(), False),
    StructField("description", StringType(), False)
])

Starting Spark application


ID,YARN Application ID,Kind,State,Spark UI,Driver log,User,Current session?
1380,application_1765289937462_1370,pyspark,idle,Link,Link,None,✔


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

## **Προεπεξεργασία Δεδομένων**
Ξεκινάμε την ανάλυση μας κάνοντας μια αρχική επεξεργασία στα δεδομένα που μας δίνονται. Η επεξεργασία αυτή αποτελείται απο την μεταφορά των δεδομένων στο περιβάλλον εργασίας μας, αλλά το σημαντικότερο την επεξεργασία των πεδίων με σκοπό την χρήση κατάλληλων τύπων.

In [2]:
### Loading of Crime data
crime_2010_2019_df = spark.read.csv(
    CRIME_DATA_2010_2019_CSV_URL,
    header=True,
    schema=RAW_CRIME_SCHEMA)

crime_2020_2025_df = spark.read.csv( 
    CRIME_DATA_2020_2025_CSV_URL,
    header=True,
    schema=RAW_CRIME_SCHEMA)

crime_df = crime_2010_2019_df.union(crime_2020_2025_df)  # Main data set to work with

### Loading Median households income by zip code
income_dy_ZIP_2021 = spark.read.csv( 
    INCOME_BY_ZIP_2021_CSV_URL,
    sep=';',
    header=True,
    schema=RAW_INCOME_BY_ZIP_2021_SCHEMA)

# convert income to numeric (remove commas/currency if present)
income_dy_ZIP_2021 = income_dy_ZIP_2021.withColumn(
    "median_income",
    expr("regexp_replace(median_income_str, '[$,]', '')").cast(IntegerType())
).filter(col("median_income").isNotNull()).drop("median_income_str")
#income_dy_ZIP_2021.show()


### Loading LA Police Station Locations
police_stations_df = spark.read.csv( 
    POLICE_STATIONS_CSV_URL,
    header=True,
    schema=POLICE_STATIONS_SCHEMA)
# police_stations_df.show()


### Loading table that contains the Race and Ethnicity code descreptions 
race_ethnicity_codes_df = spark.read.csv( 
    RACE_ETHNICITY_CODES_CSV_URL,
    header=True,
    schema=RACE_ETHNICITY_CODES_SCHEMA)
# race_ethnicity_codes_df.show()


### Loading Modus Operandi Code descreptions
raw_txt = spark.read.text(MO_CODES_TXT_URL)

mo_codes_df = (
    raw_txt
    # split only at the first space
    .withColumn("mo_code_str", split(col("value"), " ", 2)[0])
    .withColumn("description", split(col("value"), " ", 2)[1])
    # cast the code to integer (your schema expects IntegerType)
    .withColumn("mo_code", col("mo_code_str").cast("int"))
    .select("mo_code", "description")
)
del raw_txt  # discard temporary variable
# mo_code_df.show(30, truncate=False)


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

## 1. **Ανάλυση Ηλικιακών Ομάδων Θυμάτων**  
   Ταξινομούμε σε φθίνουσα σειρά τις ηλικιακές ομάδες θυμάτων που σχετίζονται με περιστατικά “βαριάς σωματικής βλάβης” (aggravated assault).  
   
   Ομάδες:  
   - Παιδιά: < 18  
   - Νεαροί ενήλικοι: 18–24  
   - Ενήλικοι: 25–64  
   - Ηλικιωμένοι: > 64

   Να υλοποιηθεί το Query 1 χρησιμοποιώντας τα DataFrame (με και χωρίς UDF) και RDD APIs.
Να εκτελέσετε και τις δύο υλοποιήσεις με 4 executors (1 core, 2GB memory). Υπάρχει διαφορά
στην επίδοση μεταξύ των τριών υλοποιήσεων;

#### **Υλοποίηση**
Παρακάτω παρατείθετε η υλοποίηση του απαραίτητου κώδικα για το ερώτημα. Στο τέλος της ενότητας υπάρχουν τα συμπεράσματα τα οποία εξάγουμε απο την εκτέλεση του.

In [3]:
spark_query_1_1 = SparkSession \
    .builder \
    .appName("Query 1: Execution with DataFrames") \
    .config("spark.executor.instances", 4) \
    .config("spark.executor.cores", 1) \
    .config("spark.executor.memory", "2g") \
    .getOrCreate()

crime_df.show(1)  # force loading of df to stabilize results

start_time = time.time()
agg_assault = crime_df.filter(crime_df.crime_code_description.contains("AGGRAVATED ASSAULT"))
grouped = agg_assault.withColumn(
    "age_group",
    when(col("victim_age") < 18, "Children")
    .when((col("victim_age") >= 18) & (col("victim_age") <= 24), "Young Adults")
    .when((col("victim_age") >= 25) & (col("victim_age") <= 64), "Adults")
    .when(col("victim_age") > 64, "Elderly")
)
result = grouped.groupBy("age_group").count()
result = result.orderBy(col("count").desc())

result.explain(mode="formatted")
result.show()

end_time = time.time()
time_df_without_udf = end_time - start_time 

print("Dataframe without UDF execution time: ", time_df_without_udf)

# clean variables
del agg_assault, grouped, result, start_time, end_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+----------------------+--------------------+--------------------+---------------+------------------+--------------------+-------------------------+-----------+----------+----------------------+--------+----------+----------+-------------------+------------+--------------------+-----------+------------------+----------------+-----------------------+------------+------------+------------+------------+----------------------+------------------+--------+---------+
|division_record_number|         report_date|     occurrence_date|occurrence_time|geographic_area_id|geographic_area_name|reporting_district_number|part_1_or_2|crime_code|crime_code_description|mo_codes|victim_age|victim_sex|victim_descent_code|premise_code| premise_description|weapon_code|weapon_description|case_status_code|case_status_description|crime_code_1|crime_code_2|crime_code_3|crime_code_4|crime_location_address|crime_cross_street|latitude|longitude|
+----------------------+--------------------+--------------------+----

In [4]:
# implementation with DataFrames (with UDF)
spark_query_1_2 = SparkSession \
    .builder \
    .appName("Query 1: Execution with DataFrames (With UDF") \
    .config("spark.executor.instances", 4) \
    .config("spark.executor.cores", 1) \
    .config("spark.executor.memory", "2g") \
    .getOrCreate()


def age_to_group(age):
    if age < 18:
        return "Childs"
    if 18 <= age <= 24:
        return "Young Adults"
    if 25 <= age <= 64:
        return "Adults"
    return "Elderly"


age_to_group_udf = udf(age_to_group, StringType())

start_time = time.time()

agg_assault = crime_df.filter(crime_df.crime_code_description.contains("AGGRAVATED ASSAULT"))

grouped = agg_assault.withColumn(
    "age_group",
    age_to_group_udf(col("victim_age"))
)

result = grouped.groupBy("age_group").count()
result = result.orderBy(col("count").desc())

result.explain(mode="formatted")
result.show()

end_time = time.time()
time_df_with_udf = end_time - start_time 

print("Dataframe with UDF execution time: ", time_df_with_udf)

# clean variables
del agg_assault, grouped, result, start_time, end_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

== Physical Plan ==
AdaptiveSparkPlan (17)
+- Sort (16)
   +- Exchange (15)
      +- HashAggregate (14)
         +- Exchange (13)
            +- HashAggregate (12)
               +- Union (11)
                  :- Project (5)
                  :  +- BatchEvalPython (4)
                  :     +- Project (3)
                  :        +- Filter (2)
                  :           +- Scan csv  (1)
                  +- Project (10)
                     +- BatchEvalPython (9)
                        +- Project (8)
                           +- Filter (7)
                              +- Scan csv  (6)


(1) Scan csv 
Output [2]: [crime_code_description#9, victim_age#11]
Batched: false
Location: InMemoryFileIndex [s3://initial-notebook-data-bucket-dblab-905418150721/project_data/LA_Crime_Data/LA_Crime_Data_2010_2019.csv]
PushedFilters: [IsNotNull(crime_code_description), StringContains(crime_code_description,AGGRAVATED ASSAULT)]
ReadSchema: struct<crime_code_description:string,victim_age:int>


In [5]:
# implementation with RDD
spark_query_1_3 = SparkSession \
    .builder \
    .appName("Query 1: Execution with RDD") \
    .config("spark.executor.instances", 4) \
    .config("spark.executor.cores", 1) \
    .config("spark.executor.memory", "2g") \
    .getOrCreate()


def age_to_group(age):
    if age < 18:
        return "Childs"
    if 18 <= age <= 24:
        return "Young Adults"
    if 25 <= age <= 64:
        return "Adults"
    return "Elderly"


crime_rdd = crime_df.rdd

start_time = time.time()

agg_assault = crime_rdd.filter(lambda row: row.crime_code_description == "AGGRAVATED ASSAULT")
age_group = agg_assault.map(lambda row: (age_to_group(row.victim_age), 1))
counts = age_group.reduceByKey(lambda a, b: a + b)
sorted_counts = counts.sortBy(lambda x: x[1], ascending=False)
sorted_counts.count()

end_time = time.time()
time_rdd = end_time - start_time

for group, count in sorted_counts.collect():
    print(f"{group}: {count}")

print("RDD execution time:", time_rdd)

# clean variables
del agg_assault, age_group, counts, sorted_counts, start_time, end_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

RDD execution time: 18.756857872009277

In [6]:
# Recap
print(
    "Execution Times:\n"
    f"{'DataFrame without UDF:':28}{time_df_without_udf:.4f} s\n"
    f"{'DataFrame with UDF:':28}{time_df_with_udf:.4f} s\n"
    f"{'RDD:':28}{time_rdd:.4f} s"
)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Execution Times:
DataFrame without UDF:      12.1898 s
DataFrame with UDF:         11.9403 s
RDD:                        18.7569 s

#### **Συμπεράσματα**
Εχοντας εκτελέσει τον απαραίτητο κώδικα για την απάντηση των queries που ζητήθηκαν, είμαστε σε θέση να εξάγουμε τα εξής συμπεράσματα, όσων αφορά την επίδοση της κάθε διαφορετικής μεθόδου εκτέλεσης. Μπορούμε να παρατηρήσουμε τον χρόνο εκτέλεσης της κάθε μεθόδου στο πίνακα στο αμέσως προηγούμενο κελί. Προκύπτει, λοιπόν,  εμφανώς, ότι την γρηγορότερη μέθοδος εκτέλεσης αποτελεί η χρήση Dataframes χωρίς  UDF, ακολουθούμενη απο την χρήση UDF, ένω η πιο αργή μέθοδος είναι η χρήση των  RDD. Αυτά τα αποτλέσματα ήταν αναμενόμενα, καθώς το Spark είναι σε θέση να δημιουργήσει, σε αυτό το απλό query τουλάχιστον, πιο αποτελεσματικές σειρές εκτέλεσης  απο αύτες που μπορεί να ορίσει ο χρήστης. Επιπλέον, σχετικά με τα RDDs, είναι γνωστό οτι είναι εγγενώς πιο αργά απο τα dataframes.

## **2. Φυλετικά Γκρουπ με τον Μεγαλύτερο Αριθμό Θυμάτων**  
   Ανά έτος βρίσκουμε τα τρία γκρουπ με τους περισσότερους καταγεγραμμένους παθόντες (Vict Descent) στο Los Angele και υπολογίζουμε το ποσοστό τους επί του συνόλου.

   Να υλοποιηθεί το Query 2 χρησιμοποιώντας τα DataFrame και SQL APIs. Να εκτελέσετε και τις
δύο υλοποιήσεις με 4 executors (1 core, 2GB memory). Να συγκρίνετε και να σχολιάσετε τους
χρόνους εκτέλεσης μεταξύ των δύο υλοποιήσεων. 

#### **Υλοποίηση**
Παρακάτω παρατείθετε η υλοποίηση του απαραίτητου κώδικα για το ερώτημα. Στο τέλος της ενότητας υπάρχουν τα συμπεράσματα τα οποία εξάγουμε απο την εκτέλεση του.

In [7]:
spark_query_2_1 = SparkSession \
    .builder \
    .appName("Query 2: Execution with DataFrames") \
    .config("spark.executor.instances", 4) \
    .config("spark.executor.cores", 1) \
    .config("spark.executor.memory", "2g") \
    .getOrCreate()

# Pre processing
#Extract year from occurrence_date (YYYY-MM-DD)
crime_df_with_year = crime_df.withColumn(
    "year",
    substring(col("occurrence_date"), 1, 4).cast("int")
)

# Keep valid victim descent entries
crime_df_with_year = crime_df_with_year.filter(col("victim_descent_code").isNotNull() & (col("victim_descent_code") != ""))


# --- Actual computation ---
start_time = time.time()

# Compute victims per (year, racial group)
year_race_df = crime_df_with_year.select('year', 'victim_descent_code')
per_race_year_counts_df = year_race_df.groupBy("year", "victim_descent_code").count()

total_victims_per_year_df = (
    per_race_year_counts_df.groupBy("year")
    .sum("count")
).withColumnRenamed("sum(count)","Total Victims Per Year")

# Join per_race_year_counts_df with yearly totals
joined_df = (
    per_race_year_counts_df.join(total_victims_per_year_df, on="year")
    .withColumn(
        "percentage",
        round(col("count") / col("Total Victims Per Year") * 100, 2)
    )
)

race_ethnicity_codes_df=race_ethnicity_codes_df.withColumnRenamed("ID", "victim_descent_code")
joined_df = joined_df.join(race_ethnicity_codes_df, on="victim_descent_code")

joined_df = joined_df.select("year", "descent", "count", "percentage")

# Select top 3 groups per year
window = Window.partitionBy("year").orderBy(col("count").desc())

result_top3_df = (
    joined_df.withColumn("rank", row_number().over(window))
    .filter(col("rank") <= 3)
    .orderBy("year", col("count").desc())
)

result_top3_df = result_top3_df.drop('rank')
result_top3_df.show(20, truncate=False)

df_execution_time = time.time() - start_time

print("Dataframe execution time:", df_execution_time)

# clean variables
del year_race_df, per_race_year_counts_df, total_victims_per_year_df,\
    joined_df, window, result_top3_df, start_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+----+----------------------+-----+----------+
|year|descent               |count|percentage|
+----+----------------------+-----+----------+
|2010|Hispanic/Latin/Mexican|73558|38.93     |
|2010|White                 |53835|28.49     |
|2010|Black                 |33937|17.96     |
|2011|Hispanic/Latin/Mexican|70845|38.8      |
|2011|White                 |51219|28.05     |
|2011|Black                 |32579|17.84     |
|2012|Hispanic/Latin/Mexican|70338|38.25     |
|2012|White                 |51839|28.19     |
|2012|Black                 |33572|18.26     |
|2013|Hispanic/Latin/Mexican|66741|37.97     |
|2013|White                 |48453|27.57     |
|2013|Black                 |31975|18.19     |
|2014|Hispanic/Latin/Mexican|68763|38.42     |
|2014|White                 |47531|26.56     |
|2014|Black                 |32952|18.41     |
|2015|Hispanic/Latin/Mexican|55978|36.65     |
|2015|White                 |44102|28.87     |
|2015|Black                 |26510|17.35     |
|2016|Hispani

In [8]:
start_time = time.time()
crime_df_with_year.createOrReplaceTempView("crime")
race_ethnicity_codes_df.createOrReplaceTempView("race")

result = spark.sql("""
WITH per_race_year_counts AS (
    SELECT
        year,
        victim_descent_code,
        COUNT(*) AS victim_count
    FROM crime
    GROUP BY year, victim_descent_code
),

totals AS (
    SELECT
        year,
        SUM(victim_count) AS total_victims_year
    FROM per_race_year_counts
    GROUP BY year
),

joined AS (
    SELECT
        yc.year,
        yc.victim_descent_code AS code,
        r.descent AS victim_descent,
        yc.victim_count,
        ROUND(yc.victim_count / yt.total_victims_year * 100, 2) AS percentage   
    FROM per_race_year_counts yc
    JOIN totals yt
        ON yc.year = yt.year
    LEFT JOIN race r
        ON yc.victim_descent_code = r.victim_descent_code
),

ranked AS (
    SELECT
        *,
        RANK() OVER (
            PARTITION BY year
            ORDER BY victim_count DESC
        ) AS rank
    FROM joined
)


SELECT
    year,
    victim_descent,
    victim_count,
    percentage
FROM ranked AS r
WHERE r.rank <= 3
ORDER BY year, victim_count DESC
""")

result.show( truncate=False)

sql_execution_time = time.time() - start_time
print("SQL execution time:", sql_execution_time)

# clean variables
del start_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+----+----------------------+------------+----------+
|year|victim_descent        |victim_count|percentage|
+----+----------------------+------------+----------+
|2010|Hispanic/Latin/Mexican|73558       |38.93     |
|2010|White                 |53835       |28.49     |
|2010|Black                 |33937       |17.96     |
|2011|Hispanic/Latin/Mexican|70845       |38.8      |
|2011|White                 |51219       |28.05     |
|2011|Black                 |32579       |17.84     |
|2012|Hispanic/Latin/Mexican|70338       |38.25     |
|2012|White                 |51839       |28.19     |
|2012|Black                 |33572       |18.26     |
|2013|Hispanic/Latin/Mexican|66741       |37.97     |
|2013|White                 |48453       |27.57     |
|2013|Black                 |31975       |18.19     |
|2014|Hispanic/Latin/Mexican|68763       |38.42     |
|2014|White                 |47531       |26.56     |
|2014|Black                 |32952       |18.41     |
|2015|Hispanic/Latin/Mexican

In [9]:
# Recap
print(
    "Execution Times:\n"
    f"{'DataFrame :':28}{df_execution_time:.4f} s\n"
    f"{'SQL API:':28}{sql_execution_time:.4f} s\n"
)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Execution Times:
DataFrame :                 6.2168 s
SQL API:                    4.8021 s

#### **Συμπεράσματα**
Τα αποτελέσματα των χρόνων εκτέλεσης φαίνονται στον παραπάνω πίνακα.

Δεν παρατηρούμε κάποια σημαντική διαφορά στον χρόνο εκτέλεσεις μεταξύ των δύο, πέρα απο ένα ισώς μικό προβάδισμα για τα SQL.

## **3. Συχνότητες Μεθόδων Διάπραξης (MOcodes)**
Ταξινομούμε τις μεθόδους διάπραξης εγκλημάτων σε φθίνουσα σειρά συχνότητας, αντιστοιχίζοντας τους κωδικούς στις περιγραφές τους.

Να υλοποιηθεί χρησιμοποιώντας τα DataFrame και RDD APIs. Να εκτελέσετε και τις δύο υλοποιήσεις με 4 executors (1 cores, 2GB memory). Να συγρκιθούν και να σχολιαστούν οι χρόνοι εκτέλεσεις για την κάθε υλοποίηση. Στη περίπτωση του DataFrame API, χρησιμοποιήστε τις μεθόδους hint και explain για να βρείτε ποιες στρατηγικές join χρησιμοποιεί ο catalyst optimizer. Πειραματιστείτε προτρέποντας το Spark να χρησιμοποιήσει διαφορετικές στρατηγικές (μεταξύ των BROADCAST, MERGE, SHUFFLE_HASH, HUFFLE_REPLICATE_NL) και σχολιάστε την επίδραση στην επίδοση. Ποιά (ή ποιές) από τις διαθέσιμες στρατηγικές join του Spark είναι καταλληλότερη(ες) και γιατί; 

#### **Υλοποίηση**
Παρακάτω παρατείθετε η υλοποίηση του απαραίτητου κώδικα για το ερώτημα. Στο τέλος της ενότητας υπάρχουν τα συμπεράσματα, τα οποία εξάγουμε απο την εκτέλεση του.

In [10]:
# implementation with dataframes
spark_query_3_1 = (
    SparkSession.builder
    .appName("MO Codes DF")
    .config("spark.executor.instances", 4)
    .config("spark.executor.cores", 1)
    .config("spark.executor.memory", "2g")
    .getOrCreate()
)

start_time = time.time()

# Count frequency of each MO code
mo_counts_df = (
    crime_df
        .filter(col("mo_codes").isNotNull())
        .groupBy("mo_codes")
        .count()
)

# Join with MO codes descriptions
joined_df = (
    mo_counts_df
        .join(mo_codes_df, mo_counts_df.mo_codes == mo_codes_df.mo_code)
        .select("mo_code", "description", "count")
        .orderBy(col("count").desc())
)

joined_df.show(20, truncate=False)

joined_df.explain(mode="formatted")

df_time = time.time() - start_time

print("DataFrame execution time:", df_time)

# clean variables
del mo_counts_df, joined_df, start_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+-------+--------------------------------------------------------------+------+
|mo_code|description                                                   |count |
+-------+--------------------------------------------------------------+------+
|344    |Removes vict property                                         |254556|
|329    |Vandalized                                                    |111449|
|1501   |Other MO (see rpt)                                            |53152 |
|325    |Took merchandise                                              |37604 |
|416    |Hit-Hit w/ weapon                                             |32571 |
|1822   |Stranger                                                      |24055 |
|377    |Takes vict's identification/driver license                    |16021 |
|400    |Force used                                                    |13454 |
|100    |Suspect Impersonate                                           |9913  |
|421    |Threaten to kill               

In [11]:
# Implementations with different hints
def run_with_hint(hint_name):
    start_time = time.time()

    # Count frequency of each MO code
    mo_counts_df = (
        crime_df
            .filter(col("mo_codes").isNotNull())
            .groupBy("mo_codes")
            .count()
    )

    # Join with MO codes descriptions
    joined = (
        mo_counts_df.hint(hint_name)
            .join(mo_codes_df, mo_counts_df.mo_codes == mo_codes_df.mo_code)
            .select("mo_code", "description", "count")
            .orderBy(col("count").desc())
    )

    joined.show(5)  # to force execution
    execution_time = time.time() - start_time

    print(f"Execution Time using {hint_name}: {execution_time:.4f} s")
    return execution_time


df_broadcast_time = run_with_hint("broadcast")
df_merge_time = run_with_hint("merge")
df_shuffle_time = run_with_hint("shuffle_hash")
df_shuffle_nl_time = run_with_hint("shuffle_hash_nl")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+-------+--------------------+------+
|mo_code|         description| count|
+-------+--------------------+------+
|    344|Removes vict prop...|254556|
|    329|          Vandalized|111449|
|   1501|  Other MO (see rpt)| 53152|
|    325|    Took merchandise| 37604|
|    416|   Hit-Hit w/ weapon| 32571|
+-------+--------------------+------+
only showing top 5 rows

Execution Time using broadcast: 4.6479 s
+-------+--------------------+------+
|mo_code|         description| count|
+-------+--------------------+------+
|    344|Removes vict prop...|254556|
|    329|          Vandalized|111449|
|   1501|  Other MO (see rpt)| 53152|
|    325|    Took merchandise| 37604|
|    416|   Hit-Hit w/ weapon| 32571|
+-------+--------------------+------+
only showing top 5 rows

Execution Time using merge: 4.9189 s
+-------+--------------------+------+
|mo_code|         description| count|
+-------+--------------------+------+
|    344|Removes vict prop...|254556|
|    329|          Vandalized|111449

In [12]:
spark_query_3_2_time = (
    SparkSession.builder
    .appName("MO Codes RDD")
    .config("spark.executor.instances", 4)
    .config("spark.executor.cores", 1)
    .config("spark.executor.memory", "2g")
    .getOrCreate()
)

start_time = time.time()

crime_rdd = crime_df.rdd
mo_rdd = mo_codes_df.rdd

# Create (mo_code, 1) pairs
pairs_rdd = (
    crime_rdd
        .filter(lambda row: row.mo_codes is not None)
        .map(lambda row: (row.mo_codes, 1))
)

# Reduce to (mo_code, count)
counts_rdd = pairs_rdd.reduceByKey(lambda a, b: a + b)

# Convert MO codes to (mo_code, description)
mo_desc_rdd = mo_rdd.map(lambda row: (row.mo_code, row.description))

# Join on mo_code
joined_rdd = counts_rdd.join(mo_desc_rdd)

# Sort decreasing by count
result_rdd = joined_rdd.sortBy(lambda x: x[1][0], ascending=False)

# Collect or show first 20
for item in result_rdd.take(20):
    print(item)

rdd_time = time.time() - start_time
print("RDD execution time:", rdd_time)


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

(344, (254556, 'Removes vict property'))
(329, (111449, 'Vandalized'))
(1501, (53152, 'Other MO (see rpt)'))
(325, (37604, 'Took merchandise'))
(416, (32571, 'Hit-Hit w/ weapon'))
(1822, (24055, 'Stranger'))
(377, (16021, "Takes vict's identification/driver license"))
(400, (13454, 'Force used'))
(100, (9913, 'Suspect Impersonate'))
(421, (7991, 'Threaten to kill'))
(2038, (7049, 'Restraining order in place between suspect and victim'))
(385, (6802, 'Suspect removed parts from vehicle'))
(2000, (6624, 'Domestic violence'))
(1402, (6378, 'Evidence Booked (any crime)'))
(913, (5708, 'Victim knew Suspect'))
(444, (5095, 'Pushed'))
(394, (4586, 'Susp takes UPS, Fedex, USPS packages'))
(930, (4430, "Unauthorized use of victim's credit/debit card or number"))
(701, (4244, 'THEFT: Trick or Device'))
(1906, (4008, 'Harrassing E-Mail/Text Message/Other Electronic Communications'))
RDD execution time: 15.578407764434814

In [13]:
# Recap
print("\nExecution times:")
print(f"DataFrame: {df_time:.4f} s")
print(f"RDD:       {rdd_time:.4f} s")
print("---------------------------")

print("\nExecution times of dataframes using different hints:")
print(f"Broadcast:                 {df_broadcast_time:.4f} s")
print(f"Merge Join:                {df_merge_time:.4f} s")
print(f"Shuffle Hash:              {df_shuffle_time:.4f} s")
print(f"Shuffle Hash Nested Loop:  {df_shuffle_nl_time:.4f} s")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…


Execution times:
DataFrame: 5.7942 s
RDD:       15.5784 s
---------------------------

Execution times of dataframes using different hints:
Broadcast:                 4.6479 s
Merge Join:                4.9189 s
Shuffle Hash:              4.5285 s
Shuffle Hash Nested Loop:  4.3364 s

#### **Συμπεράσματα**
Τα αποτελέσματα των χρόνων εκτέλεσης φαίνονται στον παραπάνω πίνακα.

Αρχικά, παρατηρούμε οτι μεταξύ του default Dataframe και του RDD, ξανά πιο γρήγορο είναι το Dataframe.
Στην συνέχεια, βλέπουμε οτι το Spark απο μόνο του χρησιμοποιεί Broadcast για το join, ενώ απο τα πειράματα μας, όπου κάνουμε force την μέθοδο που θα επιλεγεί για το join, παρατηρούμε οτι η ταχύτερη μέθοδος είναι του Shuffle Hash Nested Loop.

## **4. Απόσταση από Αστυνομικά Τμήματα**  
   Υπολογίζουμε, για κάθε τμήμα, τον αριθμό περιστατικών που βρίσκονται πιο κοντά σε αυτό από κάθε άλλο, καθώς και τη μέση απόσταση.
   
 Να υλοποιηθεί το Query 4 χρησιμοποιώντας το DataFrame ή SQL API. Για τα joins που εκτε
λούνται, να αναφέρετε ποιες στρατηγικές επιλέγει ο optimizer του Spark και να σχολιάσετε
 τις επιλογές του. Να εκτελέσετε την υλοποίησή σας εφαρμόζοντας κλιμάκωση στο σύνολο των
 υπολογιστικών πόρων που θα χρησιμοποιήσετε.

In [14]:
from pyspark.sql.functions import radians, sin, cos, acos, lit, min, asc, desc

# Implementation with DataFrame (2 executors, 1 core, 2GB memory)
spark_query_4_1_1 = (
    SparkSession.builder
    .appName("Query 4: Nearest Police Station (DataFrame)")
    .config("spark.executor.instances", 2)
    .config("spark.executor.cores", 1)
    .config("spark.executor.memory", "2g")
    .getOrCreate()
)

start_time = time.time()

# Rename columns to avoid ambiguity during join
police_renamed = police_stations_df.select(
    col("division").alias("station_division"),
    col("latitude").alias("p_lat"),
    col("longitude").alias("p_long")
)

# Remove Null Island
crime_filtered = crime_df.filter(
    (col("latitude") != 0.0) & 
    (col("longitude") != 0.0)
).select(
    "division_record_number", 
    "latitude", 
    "longitude"
)

# Every crime connects with every police station
joined_df = crime_filtered.crossJoin(police_renamed) # we use broadcast because police_renamed is short

# The Haversine formula is used to calculate the shortest distance between two points on a sphere, given their latitudes and longitudes
joined_df = joined_df.withColumn(
    "distance",
    acos(
        sin(radians(col("latitude"))) * sin(radians(col("p_lat"))) + 
        cos(radians(col("latitude"))) * cos(radians(col("p_lat"))) * cos(radians(col("p_long")) - radians(col("longitude")))
    ) * lit(6371.0)
)

# Separate the rows based on the Crime ID
window_spec = Window.partitionBy("division_record_number").orderBy(col("distance").asc())

ranked_df = joined_df.withColumn("rank", row_number().over(window_spec)) # the station with the shortest distance gets Rank 1, the second Rank 2...

# Only keep the station with the shortest distance from each crime
nearest_df = ranked_df.filter(col("rank") == 1)

# Group by Station and calculate average distance and "#"
final_result = (
    nearest_df
    .groupBy("station_division")
    .agg(
        avg("distance").alias("average_distance"),
        count("*").alias("#")
    )
    .orderBy(col("#").desc())
)

# Show results
final_result.show(truncate=False)
final_result.explain(mode="formatted")

execution_time_1 = time.time() - start_time
print("Query 4 (DataFrame) execution time (1 core, 2GB): ", execution_time_1)

del police_renamed, crime_filtered, joined_df, ranked_df, nearest_df, final_result, start_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+----------------+------------------+------+
|station_division|average_distance  |#     |
+----------------+------------------+------+
|HOLLYWOOD       |2.073572870566259 |224124|
|VAN NUYS        |2.9392526687665894|208129|
|SOUTHWEST       |2.1914657731424154|189119|
|WILSHIRE        |2.5932951164828615|186383|
|77TH STREET     |1.7171127585282697|170620|
|NORTH HOLLYWOOD |2.6424958891423715|168096|
|OLYMPIC         |1.7289297389166018|162805|
|PACIFIC         |3.8534925736593184|162027|
|CENTRAL         |0.9933230188299197|154689|
|RAMPART         |1.5342182669727789|153204|
|SOUTHEAST       |2.443911848127476 |143803|
|WEST VALLEY     |3.0215679041639985|136622|
|FOOTHILL        |4.260094410746449 |132482|
|TOPANGA         |3.296985070141389 |131054|
|HARBOR          |3.70171241411469  |127071|
|HOLLENBECK      |2.677449125015989 |116235|
|WEST LOS ANGELES|2.789517996036274 |115969|
|NEWTON          |1.6352250170121685|111392|
|NORTHEAST       |3.6229822099772404|108243|
|MISSION  

In [15]:
# Implementation with DataFrame (2 executors, 2 cores, 4GB memory)
spark_query_4_1_2 = (
    SparkSession.builder
    .appName("Query 4: Nearest Police Station (DataFrame)")
    .config("spark.executor.instances", 2)
    .config("spark.executor.cores", 2)
    .config("spark.executor.memory", "4g")
    .getOrCreate()
)

start_time = time.time()

# Rename columns to avoid ambiguity during join
police_renamed = police_stations_df.select(
    col("division").alias("station_division"),
    col("latitude").alias("p_lat"),
    col("longitude").alias("p_long")
)

# Remove Null Island
crime_filtered = crime_df.filter(
    (col("latitude") != 0.0) & 
    (col("longitude") != 0.0)
).select(
    "division_record_number", 
    "latitude", 
    "longitude"
)

# Every crime connects with every police station
joined_df = crime_filtered.crossJoin(police_renamed) # we use broadcast because police_renamed is short

# The Haversine formula is used to calculate the shortest distance between two points on a sphere, given their latitudes and longitudes
joined_df = joined_df.withColumn(
    "distance",
    acos(
        sin(radians(col("latitude"))) * sin(radians(col("p_lat"))) + 
        cos(radians(col("latitude"))) * cos(radians(col("p_lat"))) * cos(radians(col("p_long")) - radians(col("longitude")))
    ) * lit(6371.0)
)

# Separate the rows based on the Crime ID
window_spec = Window.partitionBy("division_record_number").orderBy(col("distance").asc())

ranked_df = joined_df.withColumn("rank", row_number().over(window_spec)) # the station with the shortest distance gets Rank 1, the second Rank 2...

# Only keep the station with the shortest distance from each crime
nearest_df = ranked_df.filter(col("rank") == 1)

# Group by Station and calculate average distance and "#"
final_result = (
    nearest_df
    .groupBy("station_division")
    .agg(
        avg("distance").alias("average_distance"),
        count("*").alias("#")
    )
    .orderBy(col("#").desc())
)

# Show results
final_result.show(truncate=False)

execution_time_2 = time.time() - start_time
print("Query 4 (DataFrame) execution time (2 cores, 4GB): ", execution_time_2)

del police_renamed, crime_filtered, joined_df, ranked_df, nearest_df, final_result, start_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+----------------+------------------+------+
|station_division|average_distance  |#     |
+----------------+------------------+------+
|HOLLYWOOD       |2.073572870566259 |224124|
|VAN NUYS        |2.9392526687665894|208129|
|SOUTHWEST       |2.191465773142416 |189119|
|WILSHIRE        |2.593295116482862 |186383|
|77TH STREET     |1.717112758528269 |170620|
|NORTH HOLLYWOOD |2.642495889142372 |168096|
|OLYMPIC         |1.728929738916602 |162805|
|PACIFIC         |3.853492573659318 |162027|
|CENTRAL         |0.9933230188299197|154689|
|RAMPART         |1.534218266972779 |153204|
|SOUTHEAST       |2.443911848127476 |143803|
|WEST VALLEY     |3.0215679041639985|136622|
|FOOTHILL        |4.260094410746449 |132482|
|TOPANGA         |3.2969850701413885|131054|
|HARBOR          |3.70171241411469  |127071|
|HOLLENBECK      |2.6774491250159893|116235|
|WEST LOS ANGELES|2.7895179960362735|115969|
|NEWTON          |1.6352250170121683|111392|
|NORTHEAST       |3.622982209977241 |108243|
|MISSION  

In [16]:
# Implementation with DataFrame (2 executors, 4 cores, 8GB memory)
spark_query_4_1_3 = (
    SparkSession.builder
    .appName("Query 4: Nearest Police Station (DataFrame)")
    .config("spark.executor.instances", 2)
    .config("spark.executor.cores", 4)
    .config("spark.executor.memory", "8g")
    .getOrCreate()
)

start_time = time.time()

# Rename columns to avoid ambiguity during join
police_renamed = police_stations_df.select(
    col("division").alias("station_division"),
    col("latitude").alias("p_lat"),
    col("longitude").alias("p_long")
)

# Remove Null Island
crime_filtered = crime_df.filter(
    (col("latitude") != 0.0) & 
    (col("longitude") != 0.0)
).select(
    "division_record_number", 
    "latitude", 
    "longitude"
)

# Every crime connects with every police station
joined_df = crime_filtered.crossJoin(police_renamed) # we use broadcast because police_renamed is short

# The Haversine formula is used to calculate the shortest distance between two points on a sphere, given their latitudes and longitudes
joined_df = joined_df.withColumn(
    "distance",
    acos(
        sin(radians(col("latitude"))) * sin(radians(col("p_lat"))) + 
        cos(radians(col("latitude"))) * cos(radians(col("p_lat"))) * cos(radians(col("p_long")) - radians(col("longitude")))
    ) * lit(6371.0)
)

# Separate the rows based on the Crime ID
window_spec = Window.partitionBy("division_record_number").orderBy(col("distance").asc())

ranked_df = joined_df.withColumn("rank", row_number().over(window_spec)) # the station with the shortest distance gets Rank 1, the second Rank 2...

# Only keep the station with the shortest distance from each crime
nearest_df = ranked_df.filter(col("rank") == 1)

# Group by Station and calculate average distance and "#"
final_result = (
    nearest_df
    .groupBy("station_division")
    .agg(
        avg("distance").alias("average_distance"),
        count("*").alias("#")
    )
    .orderBy(col("#").desc())
)

# Show results
final_result.show(truncate=False)

execution_time_3 = time.time() - start_time
print("Query 4 (DataFrame) execution time (4 cores, 8GB): ", execution_time_3)

del police_renamed, crime_filtered, joined_df, ranked_df, nearest_df, final_result, start_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+----------------+------------------+------+
|station_division|average_distance  |#     |
+----------------+------------------+------+
|HOLLYWOOD       |2.0735728705662595|224124|
|VAN NUYS        |2.9392526687665894|208129|
|SOUTHWEST       |2.191465773142416 |189119|
|WILSHIRE        |2.593295116482862 |186383|
|77TH STREET     |1.7171127585282695|170620|
|NORTH HOLLYWOOD |2.6424958891423715|168096|
|OLYMPIC         |1.7289297389166018|162805|
|PACIFIC         |3.853492573659316 |162027|
|CENTRAL         |0.9933230188299197|154689|
|RAMPART         |1.534218266972779 |153204|
|SOUTHEAST       |2.443911848127477 |143803|
|WEST VALLEY     |3.021567904163998 |136622|
|FOOTHILL        |4.260094410746449 |132482|
|TOPANGA         |3.2969850701413908|131054|
|HARBOR          |3.70171241411469  |127071|
|HOLLENBECK      |2.677449125015989 |116235|
|WEST LOS ANGELES|2.789517996036273 |115969|
|NEWTON          |1.6352250170121692|111392|
|NORTHEAST       |3.62298220997724  |108243|
|MISSION  

In [17]:
# Recap
print(
    "Execution Times:\n"
    f"{'1 core,  2GB memory:':28}{execution_time_1:.4f} s\n"
    f"{'2 cores, 4GB memory:':28}{execution_time_2:.4f} s\n"
    f"{'4 cores, 8GB memory:':28}{execution_time_3:.4f} s"
)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Execution Times:
1 core,  2GB memory:        21.0158 s
2 cores, 4GB memory:        15.4754 s
4 cores, 8GB memory:        14.2271 s

#### **Συμπεράσματα**

Η μετάβαση από τον 1 πυρήνα στους 2 πυρήνες επέφερε σημαντική μείωση του χρόνου, επιβεβαιώνοντας ότι **το query επωφελείται άμεσα από την παραλληλοποίηση των εργασιών**. Συγκεκριμένα, η αύξηση ως προς τους πυρήνες επέτρεψε στον executor να εκτελεί **περισσότερα tasks ταυτόχρονα** (*concurrency*), αντί για σειριακά. Παράλληλα, η μεγαλύτερη μνήμη έπαιξε καθοριστικό ρόλο, καθώς εξασφάλισε ότι τα ενδιάμεσα δεδομένα των Joins παρέμειναν στη μνήμη RAM, **αποφεύγοντας χρονοβόρες εγγραφές στον δίσκο**.

Η στρατηγική που επιλέγει ο optimizer για το crossJoin είναι η **Broadcast**, γεγονός το οποίο οφείλεται στο πολύ μικρό μέγεθος του *police_stations* (20 γραμμές). Για αυτόν τον λόγο το Spark έκρινε πιο αποδοτικό να στείλει ένα αντίγραφο του (μικρού) *police_stations* σε κάθε executor.



## **5. Συσχέτιση Εισοδήματος και Εγκληματικότητας**  
   Με βάση δεδομένα απογραφής (2020) και εισοδήματος (2021), υπολογίζουμε τη συσχέτιση μεταξύ κατακεφαλήν εισοδήματος και εγκληματικότητας ανά περιοχή.  
   Επαναλαμβάνουμε την ανάλυση για τις 10 περιοχές με το υψηλότερο και τις 10 με το χαμηλότερο εισόδημα.
   
#### **Υλοποίηση**
Παρακάτω παρατείθετε η υλοποίηση του απαραίτητου κώδικα για το ερώτημα. Στο τέλος της ενότητας υπάρχουν τα συμπεράσματα, τα οποία εξάγουμε απο την εκτέλεση του.

In [18]:
QUERY5_CONFIGS = {
    "2x4x8": {
        "executors": 2,
        "cores": 4,
        "memory": "8g"
    },
    "4x2x4": {
        "executors": 4,
        "cores": 2,
        "memory": "4g"
    },
    "8x1x2": {
        "executors": 8,
        "cores": 1,
        "memory": "2g"
    }
}

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [19]:
def run_query_5(config_key):
    config = QUERY5_CONFIGS[config_key]

    spark_query_5 = (
        SparkSession.builder
        .appName(f"Query5_{config_key}")
        .config("spark.executor.instances", config["executors"])
        .config("spark.executor.cores", config["cores"])
        .config("spark.executor.memory", config["memory"])
        .config("spark.serializer", KryoSerializer.getName)
        .config("spark.kryo.registrator", SedonaKryoRegistrator.getName)
        .config("sedona.global.index", "true")
        .getOrCreate()
    )
    
    SedonaRegistrator.registerAll(spark)
    
    # Loading GeoJSON file of LA_Census_Blocks
    blocks_df = (
        SedonaContext.create(spark_query_5).read.format("geojson")
        .option("multiLine", "true")
        .load(CENSUS_BLOCKS_2020_GEOJSON_URL)
        .selectExpr("explode(features) as features")
        .select("features.*")
    )
    
    blocks_df.show(1)  # force execution of loading
    
    start_time = time.time()  # start timing after initial loading
    
    flattened_df = blocks_df.select(
        [
         col(f"properties.{col_name}").alias(col_name)  # e.g., renames properties.POP20 to just POP20
         for col_name in blocks_df.schema["properties"].dataType.fieldNames()
        ]
        + ["geometry"]
    ).drop("properties").drop("type")
    
    census_df = (
        flattened_df
        .withColumn("POP20", col("POP20").cast(IntegerType()))
        .withColumn("ZCTA20", col("ZCTA20").cast(StringType()))
        .filter(col("COMM").isNotNull()) 
        .select("COMM", "ZCTA20", "POP20", "geometry")
    )
    
    # Inner Join: keeps rows only if the Zip Code exists in both tables
    census_income_df = census_df.join(
        income_dy_ZIP_2021, 
        census_df.ZCTA20 == income_dy_ZIP_2021.ZIP.cast("string"), 
        "inner"
    )
    
    crime_filtered = (
        crime_df
        .withColumn("year", year(to_date(col("occurrence_date"), "yyyy MMM dd hh:mm:ss a")))  # extracts only the year
        .filter((col("year") >= 2020) & (col("year") <= 2021))  # 2-year period
        .filter((col("latitude") != 0.0) & (col("longitude") != 0.0))  # remove Null Island 
        # Create Geometry Point from Lat/Lon using Sedona
        .withColumn("crime_point", expr("ST_Point(cast(longitude as Decimal(24,20)), cast(latitude as Decimal(24,20)))"))
        .select("division_record_number", "crime_point")
    )
    
    spatial_join_df = (
        crime_filtered.alias("c")
        .join(census_income_df.alias("b"), expr("ST_Contains(b.geometry, c.crime_point)"))  # does geometry hold crime_point
        .select("b.COMM", "c.division_record_number")
    )
    
    # Counting the crimes
    crime_counts_per_comm = spatial_join_df.groupBy("COMM").count().withColumnRenamed("count", "total_crimes_2yrs")
    
    comm_stats = (
        census_income_df
        .groupBy("COMM")
        .agg(
            _sum("POP20").alias("total_pop"),
            # Weighted Income = Sum(Income * Population) / Sum(Population)
            (_sum(col("median_income") * col("POP20")) / _sum("POP20")).alias("avg_income")
        )
        .filter(col("total_pop") > 0) 
    )
    
    # Money vs. Crime
    final_df = (
        comm_stats
        .join(crime_counts_per_comm, "COMM", "left")  # we keep communities even if they had zero crimes 
        .fillna(0, subset=["total_crimes_2yrs"])  # if total_crimes_2yrs = null, it means 0 crimes
        # Annual Crime Rate = (Total Crimes / 2 years) / Population
        .withColumn("crime_rate", (col("total_crimes_2yrs") / 2.0) / col("total_pop"))
        .select("COMM", "avg_income", "crime_rate")
    )

    print(f"Total Communities Analyzed: {final_df.count()}")

    corr_all = final_df.stat.corr("avg_income", "crime_rate")
    print(f"Correlation (All Areas): {corr_all:.4f}")

    top_10_rich = final_df.orderBy(desc("avg_income")).limit(10)
    corr_rich = top_10_rich.stat.corr("avg_income", "crime_rate")
    print(f"Correlation (Top 10 Richest): {corr_rich:.4f}")

    top_10_poor = final_df.orderBy(asc("avg_income")).limit(10)
    corr_poor = top_10_poor.stat.corr("avg_income", "crime_rate")
    print(f"Correlation (Top 10 Poorest): {corr_poor:.4f}")

    execution_time = time.time() - start_time
    print("Query 5 (DataFrame) execution time", config_key, ": ", execution_time)

    print("\nOptimizer Strategy for Spatial Join:")
    spatial_join_df.explain(mode="simple")

    del blocks_df, flattened_df, census_df, census_income_df
    del crime_filtered, spatial_join_df, crime_counts_per_comm, comm_stats
    del final_df, top_10_rich, top_10_poor

    return execution_time

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [20]:
execution_time_1 = run_query_5("2x4x8")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+--------------------+--------------------+-------+
|            geometry|          properties|   type|
+--------------------+--------------------+-------+
|POLYGON ((-118.48...|{1093002, 1093002...|Feature|
+--------------------+--------------------+-------+
only showing top 1 row

Total Communities Analyzed: 251
Correlation (All Areas): -0.3096
Correlation (Top 10 Richest): 0.0936
Correlation (Top 10 Poorest): -0.3701
Query 5 (DataFrame) execution time 2x4x8 :  95.1944043636322

Optimizer Strategy for Spatial Join:
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [COMM#2437, division_record_number#0]
   +- RangeJoin crime_point#2687: geometry, geometry#2405: geometry, WITHIN
      :- Union
      :  :- Project [division_record_number#0,  **org.apache.spark.sql.sedona_sql.expressions.ST_Point**   AS crime_point#2687]
      :  :  +- Filter ((((((isnotnull(latitude#26) AND isnotnull(longitude#27)) AND (year(cast(gettimestamp(occurrence_date#2, yyyy MMM dd hh:mm:ss a, Ti

In [21]:
execution_time_2 = run_query_5("4x2x4")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+--------------------+--------------------+-------+
|            geometry|          properties|   type|
+--------------------+--------------------+-------+
|POLYGON ((-118.48...|{1093002, 1093002...|Feature|
+--------------------+--------------------+-------+
only showing top 1 row

Total Communities Analyzed: 251
Correlation (All Areas): -0.3096
Correlation (Top 10 Richest): 0.0936
Correlation (Top 10 Poorest): -0.3701
Query 5 (DataFrame) execution time 4x2x4 :  72.62918758392334

Optimizer Strategy for Spatial Join:
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [COMM#3712, division_record_number#0]
   +- RangeJoin crime_point#3962: geometry, geometry#3680: geometry, WITHIN
      :- Union
      :  :- Project [division_record_number#0,  **org.apache.spark.sql.sedona_sql.expressions.ST_Point**   AS crime_point#3962]
      :  :  +- Filter ((((((isnotnull(latitude#26) AND isnotnull(longitude#27)) AND (year(cast(gettimestamp(occurrence_date#2, yyyy MMM dd hh:mm:ss a, T

In [22]:
execution_time_3 = run_query_5("8x1x2")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+--------------------+--------------------+-------+
|            geometry|          properties|   type|
+--------------------+--------------------+-------+
|POLYGON ((-118.48...|{1093002, 1093002...|Feature|
+--------------------+--------------------+-------+
only showing top 1 row

Total Communities Analyzed: 251
Correlation (All Areas): -0.3096
Correlation (Top 10 Richest): 0.0936
Correlation (Top 10 Poorest): -0.3701
Query 5 (DataFrame) execution time 8x1x2 :  69.15761709213257

Optimizer Strategy for Spatial Join:
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [COMM#4987, division_record_number#0]
   +- RangeJoin crime_point#5237: geometry, geometry#4955: geometry, WITHIN
      :- Union
      :  :- Project [division_record_number#0,  **org.apache.spark.sql.sedona_sql.expressions.ST_Point**   AS crime_point#5237]
      :  :  +- Filter ((((((isnotnull(latitude#26) AND isnotnull(longitude#27)) AND (year(cast(gettimestamp(occurrence_date#2, yyyy MMM dd hh:mm:ss a, T

In [23]:
print("Execution Times")
print("-" * 45)
print(f"{'2 executors, 4 cores, 8GB':30}{execution_time_1:.4f} s")
print(f"{'4 executors, 2 cores, 4GB':30}{execution_time_2:.4f} s")
print(f"{'8 executors, 1 core, 2GB':30}{execution_time_3:.4f} s")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Execution Times
---------------------------------------------
2 executors, 4 cores, 8GB     95.1944 s
4 executors, 2 cores, 4GB     72.6292 s
8 executors, 1 core, 2GB      69.1576 s

#### **Συμπεράσματα**

* Όσον αφορά το execution time, διαπιστώνουμε ότι **η υλοποίηση με 8 executors × 1 core, 2GB memory είναι η γρηγορότερη, ενώ αυτή με 2 executors × 4 cores, 8GB memory παίρνει τον περισσότερο χρόνο**. Τα αποτελέσματα αυτά συνάδουν απόλυτα με το γεγονός πως **ένα Spatial Join επωφελείται περισσότερο από περισσότερους executors, παρά από περισσότερους cores**. Με αυτόν τον τρόπο πιθανώς επιτυγχάνεται καλύτερη παραλληλοποίηση. 

* Το *correlation number* είναι ένας αριθμός ο οποίος μας δείχνει την σχέση μεταξύ δύο μεγεθών ("avg_income", "crime_rate"). Θετικός αριθμός θα προκύψει άμα και τα δύο μεγέθη αυξάνονται (ή μειώνονται) μαζί, ενώ σε αντίθετη περίπτωση, ένας αρνητικός αριθμός προδίδει την ύπαρξη μιας "αντίστροφης" σχέσης μεταξύ των μεγεθών. Προφανώς, άμα το correlation number είναι ίσο με μηδέν, αυτό σημαίνει ότι ο ένας παράγοντας δεν επηρεάζει με κανέναν τρόπο τον άλλον. Τα αποτελέσματα της υλοποίησης συμφωνούν απόλυτα με αυτή τη λογική:

  • **Correlation (All Areas): -0.3096** : Αυτός ο αριθμός μας δείχνει ότι σε γενικές γραμμές, η οικονομική σταθερότητα συνδέεται με την μείωση του εγκλήματος. Είναι κάτι το οποίο βγάζει απόλυτο νόημα και στον πραγματικό κόσμο, καθώς υψηλότερο εισόδημα συνεπάγεται συνήθως μια πιο ασφαλής περιόχη, αν και όχι άτρωτη από εγκλήματα. Ο αριθμός δε βρίσκεται κοντά στο -1, οπότε αυτό σημαίνει ότι το εισόδημα δεν είναι ο μόνος παράγοντας που επηρεάζει την εγκληματική δραστηριότητα. 

  • **Correlation (Top 10 Richest): 0.0936** : Για τις δέκα πιο πλούσιες περιοχές, ο αριθμός βρίσκεται πολύ κοντά στο μηδέν: αυτό σημαίνει ότι το εισόδημα σταματάει να παίζει μεγάλο ρόλο στο επίπεδο ασφάλειας. Τα περισσότερα χρήματα δεν ισοδυναμούν αναγκαστικά με λιγότερα εγκλήματα, αλλά η σχέση που τα συνδέει γίνεται πιο ασθενής.

  • **Correlation (Top 10 Poorest): -0.3701** : Στις δέκα πιο φτωχές περιοχές, η εγκληματικότητα και το εισόδημα συνδέονται με άμεσο τρόπο. Ακόμη και οι πιο μικρές αλλαγές στην οικονομική κατάσταση μιας περιοχής μπορεί να έχουν σημαντική επίδραση στον βαθμό εγκληματικότητας, με την έλλειψη σταθερότητας να παίζει σημαντικό ρόλο σε αυτό. Το νούμερο αυτό υποδεικνύει πόσο σημαντική είναι η ανάπτυξη τέτοιων περιοχών χαμηλού εισοδήματος, καθώς ακόμη και η πιο μικρή επένδυση θα μπορούσε να βοηθήσει τη μείωση του εγκλήματος.


* Τέλος, όσον αφορά τη στρατηγική join που χρησιμοποιείται, στρατηγικές όπως οι *SortMergeJoin* και *ShuffleHashJoin* δε υποστηρίζουν συνθήκες που δεν είναι ισότητα. Για το Spatial Join, η στρατηγική που επιλέχθηκε είναι η **RangeJoin**, η οποία έχει ειδικά σχεδιαστεί για να χειρίζεται ζητήματα με *ST_Contains*. Η *RangeJoin* ανιχνεύει σημεία μέσα σε πολύγωνα χωρίς να αυξήσει την ακρίβεια του προγράμματος παραπάνω από το φυσιολογικό, καθιστώντας την ιδανική για ένα γεωμετρικό πρόβλημα. Αξίζει να σημειωθεί, επίσης, ότι ο optimizer εντόπισε πως το dataset με τα εισοδήματα είναι μάλλον σχετικά μικρό, επιλέγοντας έτσι τη στρατηγική **BroadcastHashJoin** για να συνδέσει προηγουμένως τα δεδομένα του εισοδήματος με τα Census Βlocks. 